### ISSUE TITLE
[Feature] Install/Update bundle directly from IDE to remote OSGi runtimes

### ISSUE THREAD

**Issue by amitjoy** [bndtools/bnd#4826]
We can introduce a feature which would enable us to remotely install or update a bundle from IDE directly. We can introduce a menu item which will only be activated after selecting a jar and we can select the remote IP where we want to install/update the bundle. The IPs of remote runtimes can also be stored in Eclipse preferences for continuous usage. 

I am currently starting to work on it.

**Comment by stbischof** [comment]
You are speaking about `remote agent`?
https://bnd.bndtools.org/chapters/300-launching.html
https://www.datainmotion.de/blog-bnd-remote-launching/

Would be nice!

**Comment by amitjoy** [comment]
Yeah, exactly. The feature I am proposing would leverage the bnd remote agent but more in UI friendly manner where one can select one of your previously stored remote OSGi runtimes (associated with IPs) to install a jar by selecting it from `Bndtools Explorer` and clicking on `Install to` menu option.

**Comment by kriegfrj** [comment]
This sounds like a really useful feature (although I've never used the remote agent myself).

May I strongly suggest that:
1. You create a separate bundle for this new feature rather than putting it in `bndtools.core` (eg, `org.bndtools.remote` or similar);
2. You use the "facade" pattern that we have implemented in `org.bndtools.launch`. I have an incomplete rough draft of a howto here: https://github.com/kriegfrj/bnd/blob/0a584d024b783ed5df1c0ad2e8eabe30beb5713b/DEV_README.md You can also look in the `org.bndtools.launch` bundle for an example.

**Comment by amitjoy** [comment]
@kriegfrj Thanks a lot for your valuable feedback. Yesterday, I started with the similar approach to make UI related stuffs more modular. Therefore, I created a separate project - `bndtools.remoteinstall`. It makes more sense to create own project as there would be quite a few classes.

I would carefully go through your draft regarding `ExtensionFacade` as well and try to integrate where necessary.

**Comment by kriegfrj** [comment]
Just FYI, @amitjoy - I have finished that draft now. Read `DEV_README.md` for a description on how to use the `ExtensionFacade`.

Feedback welcomed! :smile:

**Comment by amitjoy** [comment]
@kriegfrj Thanks a lot for the detailed info regarding `ExtensionFacade`. It really makes development a lot easier. I really liked it. During some testing, I noticed that not all extension points are supported by `ExtensionFacade`, for example, `<dynamic class="X">` for `ContributionItem` isn't supported as Eclipse throws an exception saying that `Expected ContributionItem, found proxy`. That's why, I had to give up using dynamic menu items. However, I feel, the advantages `ExtensionFacade` brings outweigh the drawbacks in all aspects.

**Comment by kriegfrj** [comment]
Yeah unfortunately (as the current draft notes) there are cases where Eclipse can make it hard. Did you try using the factory mode?

An enhancement (for another day) would be to use Byte Buddy to bender the proxies so that it could work with classes and not only interfaces.  That might also fix your problem with ContributionItem.

**Comment by amitjoy** [comment]
@kriegfrj Yeah, I thought the same that `ByteBuddy` might fix the problem but for now, it is not necessary as I don't need the dynamic `ContributionItem` anymore. We can, later, introduce it in `ExtensionFacade`.
